runtime/vm/resolvefield.cpp

```diff
@@ -578,6 +583,124 @@ addHiddenInstanceField(J9JavaVM *vm, const char *className, const char *fieldNam
 	return 0;
 }
 
+/**
+ * Report a hot field if the JIT has determined that the field has met appropriate thresholds to be determined a hot field. 
+ * Valid if dynamicBreadthFirstScanOrdering is enabled.
+ *
+ * @param javaVM[in] pointer to the J9JavaVM
+ * @param reducedCpuUtil normalized cpu utilization of the method reporting the hot field
+ * @param clazz pointer to the class where a hot field should be added/updated
+ * @param fieldOffset value of the field offset that should be added/updated as a hot field for the given class
+ * @param reducedFrequency normalized block frequency of the hot field for the method reporting the hot field
+ */
+void
+reportHotField(J9JavaVM *javaVM, int32_t reducedCpuUtil, J9Class* clazz, U_8 fieldOffset,  U_32 reducedFrequency)
+{
+	/* exit if pools are null */
+	if (NULL == clazz->classLoader->hotFieldPool || NULL == javaVM->hotFieldClassInfoPool) {
+		return;
+	}
+
+	/* If hotFieldsInfo pool element for the class does not exist already, create and initialize the class' hotFieldsInfo pool element
+	 * otherwise, create/update the given hot field for the class 
+	 */
+	if (NULL == clazz->hotFieldsInfo) {
+		createClassHotFieldsInfo(javaVM, clazz, fieldOffset, reducedCpuUtil, reducedFrequency);
+	} else if (clazz->hotFieldsInfo->hotFieldListLength < MAX_HOT_FIELD_LIST_LENGTH) {
+		addOrUpdateHotField(clazz, fieldOffset, reducedCpuUtil, reducedFrequency);
+	}
+}
+
+/**
+ * Create and initialize a class' hotFieldsInfo pool element.
+ * Valid if dynamicBreadthFirstScanOrdering is enabled.
+ *
+ * @param javaVM[in] pointer to the J9JavaVM
+ * @param clazz pointer to the class where a hot field should be added
+ * @param fieldOffset value of the field offset that should be added as a hot field for the given class
+ * @param reducedCpuUtil normalized cpu utilization of the method reporting the hot field
+ * @param reducedFrequency normalized block frequency of the hot field for the method reporting the hot field
+ */
+VMINLINE void
+createClassHotFieldsInfo(J9JavaVM *javaVM, J9Class* clazz, U_8 fieldOffset, int32_t reducedCpuUtil, U_32 reducedFrequency)
+{
+	omrthread_monitor_enter(javaVM->hotFieldClassInfoPoolMutex);
+	if (NULL == clazz->hotFieldsInfo) {
+		/* initialize new hotFieldsInfo pool element as it does not exist already */
+		clazz->hotFieldsInfo = (J9ClassHotFieldsInfo *)pool_newElement(javaVM->hotFieldClassInfoPool);
+		if (NULL == clazz->hotFieldsInfo) {
+			goto exit;
+		}
+		clazz->hotFieldsInfo->hotFieldListLength = 0;
+		clazz->hotFieldsInfo->consecutiveHotFieldSelections = 0;
+		clazz->hotFieldsInfo->hotFieldOffset1 = U_8_MAX;
+		clazz->hotFieldsInfo->hotFieldOffset2 = U_8_MAX;
+		
+		/* create and initialize the new hot field */
+		omrthread_monitor_enter(clazz->classLoader->hotFieldPoolMutex);
```

Comment:
same thing here, I would remove this code. Just make sure `addOrUpdateHotField` is called after

Reply:
The issue with doing this would be that the `addOrUpdateHotField ` method does not handle setting the `clazz->hotFieldsInfo->hotFieldListHead`. If `addOrUpdateHotField` was modified to handle the setting of the hotfield list head, then the `hotFieldClassInfoPoolMutex` would need to be acquired in `addOrUpdateHotField` method as well to handle this case which would add lock contention. Is this desired over this code duplication?

Reply:
Strictly speaking, you only need the `hotFieldClassInfoPoolMutex`  for allocating the  `hotFieldsInfo`. As long as `hotFieldListHead` is only ever updated while `hotFieldPoolMutex` is held then there is no problem setting `hotFieldListHead` in `addOrUpdateHotField` without acquiring `hotFieldClassInfoPoolMutex`.